### PR TITLE
Add function 'age'

### LINKED ISSUE #41115: age function (unit: week /  month  / quarter / year  )
a function should return the number of the full units ( age in years, ...)

```sql
SELECT
    toDate('2022-01-01') AS e,
    toDate('2021-12-29') AS s,
    dateDiff('month', s, e) AS month__diff,
    dateDiff('year', s, e) AS year_diff

┌──────────e─┬──────────s─┬─month__diff─┬─year_diff─┐
│ 2022-01-01 │ 2021-12-29 │           1 │         1 │
└────────────┴────────────┴─────────────┴───────────┘
```

```sql
SELECT
    toDate('2022-01-01') AS e,
    toDate('2021-12-29') AS s,
    age('month', s, e) AS month__diff,
    age('year', s, e) AS year_diff
┌──────────e─┬──────────s─┬─month__diff─┬─year_diff─┐
│ 2022-01-01 │ 2021-12-29 │           0 │         0 │
└────────────┴────────────┴─────────────┴───────────┘
```

```sql
CREATE FUNCTION age AS (e, s) -> (dateDiff('year', s, e) - ((month(s), day(s)) > (month(e), day(e))))

SELECT age(toDate('2022-01-01'), toDate('2021-12-29')) AS age

┌─age─┐
│   0 │
└─────┘
```

### PR DESCRIPTION
**Pull request by rvasin**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added age function to calculate difference  between two dates or dates with time values expressed as number of full units. 
Close #41115

### CODE REVIEW

**Comment by rvasin** [comment]
@rschu1ze Please be reviewer of this PR and assign "can be tested" label.

**Comment by rschu1ze** [comment]
@roman I'll try to take a look but can't promise it because I'll be on vacation in the next two weeks.

**Comment by rvasin** [comment]
OK. Thank you. Have a nice vocation! 

There are several point to note about this PR. One point: in some of other DBMS like PostgreSQL usually first argument is end date and second is start date. But in CH `dateDiff` function it's in reverse sense: first (the first after unit argument) is start date and second is end date. Thus in `age` function I tried to keep the same logic: first is start date and second is end date. If needed we may change it (to make it compatible with other DBMS). But other DBMS usually don't have unit parameter.

File: tests/queries/0_stateless/02477_age.sql
Comment (rschu1ze) [inline tests/queries/0_stateless/02477_age.sql:14]: I am confused. On my system and in the expected results,

```sql
SELECT age('month', toDate('2017-12-31'), toDate('2018-01-01'));
-- 0
```

but

```sql
SELECT age('week', toDate('2017-12-31'), toDate('2018-01-01'));
-- 1
```

Is that because of funny time zone effects?

File: src/Functions/dateDiff.cpp
Comment (rschu1ze) [inline src/Functions/dateDiff.cpp:272]: How about template parameter "is_absolute" or "is_relative"?

**Review by rschu1ze — COMMENTED** [review]
I am back from vacation. Thanks for the PR, please find some comments attached.

File: src/Functions/dateDiff.cpp
Comment (rvasin) [reply src/Functions/dateDiff.cpp:452]: Hi, Robert, Nice to hear you again! I will do the work on the suggested changes tomorrow. So you suggest to remove the docs for age() function, so make it this way:
```
REGISTER_FUNCTION(Age)
{
    factory.registerFunction<FunctionDateDiff<false>>({}, FunctionFactory::CaseInsensitive);
}
```
and then add the function name into tests/queries/0_stateless/02415_all_new_functions_must_be_documented.sql and corresponding .reference file? Please confirm.

Comment (rschu1ze) [reply src/Functions/dateDiff.cpp:452]: Confirmed :)

Comment (rvasin) [reply src/Functions/dateDiff.cpp:452]: Done

File: src/Functions/dateDiff.cpp
Comment (rvasin) [reply src/Functions/dateDiff.cpp:232]: Done. Thank you for noticing it.

File: tests/queries/0_stateless/02477_age.sql
Comment (rvasin) [reply tests/queries/0_stateless/02477_age.sql:14]: I am confused too. Because I get 0 when I run the query form clickhouse-client and 1 then the query is run from test. I am working on it.

File: src/Functions/TransformDateTime64.h
Comment (rvasin) [reply src/Functions/TransformDateTime64.h:51]: Done

File: tests/queries/0_stateless/02477_age.sql
Comment (rvasin) [reply tests/queries/0_stateless/02477_age.sql:48]: OK. Done

File: tests/queries/0_stateless/02477_age.sql
Comment (rvasin) [reply tests/queries/0_stateless/02477_age.sql:82]: Why this one is not `age`?

File: docs/en/sql-reference/functions/date-time-functions.md
Comment (rvasin) [reply docs/en/sql-reference/functions/date-time-functions.md:649]: `age` is not a good name, but other DBMS (while not all) use it:
1. [PostgreSQL](https://www.w3resource.com/PostgreSQL/age-function.php )
2. [IBM DB2](https://www.ibm.com/docs/en/db2/11.5?topic=functions-age)
3. Vertica: [AGE_IN_YEARS](https://www.vertica.com/docs/9.2.x/HTML/Content/Authoring/SQLReferenceManual/Functions/Date-Time/AGE_IN_YEARS.htm ) and [AGE_IN_MONTHS](https://www.vertica.com/docs/9.2.x/HTML/Content/Authoring/SQLReferenceManual/Functions/Date-Time/AGE_IN_MONTHS.htm)

File: tests/queries/0_stateless/02477_age_via_unix_epoch.sql
Comment (rvasin) [reply tests/queries/0_stateless/02477_age_via_unix_epoch.sql:1]: There was a problem when `dateDiff` worked well with dates before and after Unix epoch. Please read #41978 So I fixed it in `dateDiff` function and created a test "via_unix_epoch" specially for that.
Maybe for `age` function such test is not needed. Because at least it's converted by test of `dateDiff`.  I may remove the test for `age` function.

File: tests/queries/0_stateless/02477_age.sql
Comment (rschu1ze) [reply tests/queries/0_stateless/02477_age.sql:82]: I looked into the wrong place, sorry.

File: docs/en/sql-reference/functions/date-time-functions.md
Comment (rschu1ze) [reply docs/en/sql-reference/functions/date-time-functions.md:649]: Hm, okay, I didn't google hard enough :)

File: tests/queries/0_stateless/02477_age.sql
Comment (rvasin) [reply tests/queries/0_stateless/02477_age.sql:14]: Found the issue and corrected it. I will push my changes later on today.

File: tests/queries/0_stateless/02477_age_via_unix_epoch.sql
Comment (rvasin) [reply tests/queries/0_stateless/02477_age_via_unix_epoch.sql:1]: Removed 02477_age_via_unix_epoch test

File: docs/en/sql-reference/functions/date-time-functions.md
Comment (rvasin) [reply docs/en/sql-reference/functions/date-time-functions.md:640]: Done both points ("date or date with time" and moving).

File: docs/en/sql-reference/functions/date-time-functions.md
Comment (rvasin) [reply docs/en/sql-reference/functions/date-time-functions.md:642]: Done. Please verify

File: docs/en/sql-reference/functions/date-time-functions.md
Comment (rvasin) [reply docs/en/sql-reference/functions/date-time-functions.md:643]: done

File: src/Functions/dateDiff.cpp
Comment (rvasin) [reply src/Functions/dateDiff.cpp:272]: Done. Replaced it with "is_relative"

File: docs/en/sql-reference/functions/date-time-functions.md
Comment (rschu1ze) [reply docs/en/sql-reference/functions/date-time-functions.md]: Lgtm.

File: tests/queries/0_stateless/02477_age.sql
Comment (rschu1ze) [inline tests/queries/0_stateless/02477_age.sql:4]: The "negative" calculations (1st argument > 2nd argument) are a bit weird. E.g. is it really expected that the year component of the absolute difference between 01-01 and 12-31 is -1 years? It is more like 0.99999 years = 0 years.

**Review by rschu1ze — COMMENTED** [review]
Looks good to me. Maybe you could clarify if results for startdate > enddate work as expected - thanks. After that we can merge.

File: tests/queries/0_stateless/02477_age.sql
Comment (rvasin) [reply tests/queries/0_stateless/02477_age.sql:4]: I corrected it and pushed my changes. Please take a look.

To compare startdate > enddate we cannot rely on `res` like:
`if (res < 0)`
because in some cases res ==0 for both for startdate > enddate and for startdate < enddate

`res` is calculated as here:
```
 auto res = static_cast<Int64>(transform_y.execute(y, timezone_y))
                - static_cast<Int64>(transform_x.execute(x, timezone_x));
```
So I had to calculate seconds first and then compare the seconds. 
```
            auto x_seconds = TransformDateTime64<ToRelativeSecondNumImpl<ResultPrecision::Extended>>(transform_x.getScaleMultiplier()).execute(x, timezone_x);
            auto y_seconds = TransformDateTime64<ToRelativeSecondNumImpl<ResultPrecision::Extended>>(transform_y.getScaleMultiplier()).execute(y, timezone_y);
            if (x_seconds <= y_seconds)

```
Because as far as I know there are no  DateTimeTransforms to compare  different types of x and y. For example when `x` is `Date32` and `y` is `DateTime64`.

I also renamed `x_comp` to `a_comp` and `y_comp` to `b_comp` to avoid confusion what is `x` or `y` in later comparisons.

Comment (rschu1ze) [reply tests/queries/0_stateless/02477_age.sql:4]: Looks good to me, thanks!

**Review by rschu1ze — APPROVED** [review]

**Comment by rschu1ze** [comment]
Will merge once builds/tests finished.

File: tests/queries/0_stateless/02477_age.sql
Comment (rvasin) [reply tests/queries/0_stateless/02477_age.sql:4]: Great. Thank you for reviewing it!

**Comment by rschu1ze** [comment]
I don't really understand the error in [Docs Check](https://s3.amazonaws.com/clickhouse-test-reports/43123/620df1ec6f7f5014b2b44210f063127a7f861706/docs_check/runlog.log), perhaps @DanRoscigno has an idea?

**Comment by rvasin** [comment]
> I don't really understand the error in [Docs Check](https://s3.amazonaws.com/clickhouse-test-reports/43123/620df1ec6f7f5014b2b44210f063127a7f861706/docs_check.html) 

I have also noticed this error. It's related to size/formatting of Changelog 2019.

I think I merged "unlucky" version of master into my feature branch yesterday. So I updated again my branch with master to see how it will go now.

File: src/Functions/DateTimeTransforms.h
Comment (rschu1ze) [inline src/Functions/DateTimeTransforms.h:1346]: I just found out about #43466 and as far as I see, a similar fix (`DateLUTImpl::Time` instead of `time_t`, l. 1356) is needed.

Comment (rvasin) [reply src/Functions/DateTimeTransforms.h:1346]: Thank you for noticing it. I will update it in my branch in about 3 hours from now.

Comment (rvasin) [reply src/Functions/DateTimeTransforms.h:1346]: I replaced time_t with DateLUTImpl::Time and pushed my changes. It seems it's OK now with Docs Check?

**Comment by rschu1ze** [comment]
Looks good to me, but you would need to merge master into this PR once more (sorry, I don't have the rights) to get fix #43911 for the broken clang-tidy build. Right now, the merge button is greyed out for me.

**Comment by rvasin** [comment]
> Looks good to me, but you would need to merge master into this PR once more (sorry, I don't have the rights) to get fix #43911 for the broken clang-tidy build. Right now, the merge button is greyed out for me.

I pressed Update button to merge the master branch into my feature branch.

**Comment by alexey-milovidov** [comment]
This pull request introduced a regression on Mac OS.

Most likely, it's related to the fact that:
`int64_t` is `long` on Linux
`int64_t` is `long long` on Mac OS
The `long` and `long long` types have identical behavior (the only exception is Windows) but are distinguished by the compiler for the purpose of overload resolution.

**Comment by alexey-milovidov** [comment]
```
:) select 1670853969::DateTime

SELECT CAST('1670853969', 'DateTime')

Query id: 14533336-f2e9-41ad-aefa-9d5084c4119d

┌─CAST('1670853969', 'DateTime')─┐
│            2022-12-12 14:06:09 │
└────────────────────────────────┘
```

**Comment by rvasin** [comment]
@alexey-milovidov Please explain how CAST() is related to dateDiff() or age() functions? 
It this PR I tried to not touch much outside of dateDiff.cpp.

Are you sure that it's because of this PR? So after revert of this PR CAST() function works fine now on Mac OS? Sorry, I don't have Mac OS so I cannot test it.

**Comment by rschu1ze** [comment]
This PR was determined as culprit by bisecting on Mac done internally. I haven't had time to reproduce/check yet but will do hopefully soon.

**Comment by rvasin** [comment]
OK. Thank you. I though about bisecting.

I will try to run CAST function in debugger on Monday to see which functions does it call. 

What I added in this PR is:

```
    TransformDateTime64(DateTime64::NativeType scale_multiplier_ = 1) /// NOLINT(google-explicit-constructor)
        : scale_multiplier(scale_multiplier_)
    {}
```

While before my PR there was:

```
    // non-explicit constructor to allow creating from scale value (or with no scale at all), indispensable in some contexts.
    TransformDateTime64(UInt32 scale_ = 0) /// NOLINT
        : scale_multiplier(DecimalUtils::scaleMultiplier<DateTime64::NativeType>(scale_))
    {}
```

Besides that I don't see other suspicions places right now which may affect CAST(). It's a pity I don't have Mac OS to test it in live.

**Comment by rschu1ze** [comment]
I can reproduce the issue locally on Mac (after some messing with libunwind).

So far it looks like method `toDateTimeComponents()` in Common/DateLUTImpl.h is the culprit. It seems that a wrong overload gets chosen on Mac.

**Comment by rvasin** [comment]
Robert, yes. I see that for the call:

`select 1670853969::DateTime`

It calls:
src/Common/LocalDateTime.h
```
    void init(time_t time, const DateLUTImpl & time_zone)
    {
        DateLUTImpl::DateTimeComponents components = time_zone.toDateTimeComponents(time);
```

BTW toDateTimeComponents is used in just 3 places:
first is listed above and 2nd and 3rd:

```
    inline UInt64 toNumYYYYMMDDhhmmss(Time t) const
    {
        DateTimeComponents components = toDateTimeComponents(t);
```

```
    inline std::string timeToString(Time t) const
    {
        DateTimeComponents components = toDateTimeComponents(t);
```
Note: in first place there is `time_t` as for argument of toDateTimeComponents
and in 2nd and 3rd places there is `Time`.

Do you remember you wrote:

> I just found out about https://github.com/ClickHouse/ClickHouse/pull/43466 and as far as I see, a similar fix (DateLUTImpl::Time instead of time_t, l. 1356) is needed.

The issue was also Mac OS related. 

So maybe we just need to replace `time_t` by `Time`? Like this:

```
    void init(Time time, const DateLUTImpl & time_zone)
    {
        DateLUTImpl::DateTimeComponents components = time_zone.toDateTimeComponents(time);
```

Could you please try it on Mac OS?

**Comment by rschu1ze** [comment]
Thanks, Roman. I fixed the issue similar how you proposed in [this revert-revert](https://github.com/ClickHouse/ClickHouse/pull/44421). Will also check if there are more places where `time_t` is used but `Time` should really be used.

**Pull request merged**
